addons/stock/models/stock_move.py

```diff
@@ -815,31 +809,84 @@ def onchange_date(self):
+                else:
+                    self.update({'move_line_nosuggest_ids': move_lines_commands})
+                break
+
+    def _generate_serial_move_line_commands(self, lot_names, origin_move_line=None):
+        """Return a list of commands to update the move lines (write on
+        existing one or create new one).
+        Called when user want to create and assign multiple serial numbers in
+        one time (using the button/wizard or copy/paste a list in the field).
+
+        :param lot_names: A list containing all serial number to assign.
+        :type lot_names: list
+        :param origin_move_line: A move line to duplicate the value from, default to None
+        :type origin_move_line: record of :class:`stock.move.line`
+        :return: A list of commands to create/update :class:`stock.move.line`
+        :rtype: list
+        """
+        self.ensure_one()
+
+        def without_lot_filter(move_line):
+            return not move_line.lot_id and not move_line.lot_name
+
+        # Select the right move lines depending of the picking type configuration.
+        move_lines = self.env['stock.move.line']
+        if self.picking_type_id.show_reserved:
+            move_lines = self.move_line_ids.filtered(without_lot_filter)
+        else:
+            move_lines = self.move_line_nosuggest_ids.filtered(without_lot_filter)
+        move_lines_commands = []
+        for lot_name in lot_names:
+            # We write the lot name on an existing move line (if we have still one)...
+            if len(move_lines):
+                move_lines_commands.append((1, move_lines[0].id, {
+                    'lot_name': lot_name,
+                    'qty_done': 1,
+                }))
+                move_lines = move_lines[1:]
+            # ... or create a new move line with the serial name.
+            else:
+                # New move line are based on the one we passed if we passed one,
+                # otherwise it will base on the move.
+                record = origin_move_line or self
+                location_dest = record.location_dest_id._get_putaway_strategy(record.product_id)
+
+                move_line_vals = {
+                    'lot_name': lot_name,
+                    'qty_done': 1,
+                    'product_id': record.product_id.id,
+                    'product_uom_id': record.product_id.uom_id.id,
+                    'location_id': record.location_id.id,
+                    'location_dest_id': location_dest.id or record.location_dest_id.id,
+                    # `owner_id` and `package_id` are take only in the case we
+                    # create a new move line from an existing move line.
+                    'owner_id': origin_move_line and origin_move_line.owner_id.id,
+                    'package_id': origin_move_line and origin_move_line.package_id.id,
+                }
+                move_lines_commands.append((0, 0, move_line_vals))
+        return move_lines_commands
```

Comment:
please don't put this helper between two onchanges methods

1. Private attributes (_name, _description, _inherit, …)
2. Default method and _default_get
3. Field declarations
4. Compute, inverse and search methods in the same order as field declaration
5. Selection method (methods used to return computed values for selection fields)
6. Constrains methods (@api.constrains) and onchange methods (@api.onchange)
7. CRUD methods (ORM overrides)
8. Action methods
9. And finally, other business methods.

i know it'a mess at the moment but we should try to add new methds where they should be